File: 백준/Silver/2108. 통계학/통계학.py
import sys

num_list = []

N = int(sys.stdin.readline())
for _ in range(N):
    num = int(sys.stdin.readline())
    num_list.append(num)

num_list.sort()
# 산술평균
def avg(list):
    return sum(list)/len(list)
# 중앙값
def mid(list):
    return list[len(list)//2]
# 최빈값
def mode(list):
    #받는 리스트가 한개의 원소일 경우 그냥 바로 그 값 출력
    if len(list)==1:
        return list[0]

    mode_list = [] # 최빈값들의 리스트
    most = 0 # 최빈값 반복 횟수
    n = 0 # 반복 횟수
    for i in range(1,len(list)):
        if list[i]==list[i-1]:
            n += 1
            if n>most:
                most = n
                mode_list.clear() # 기존의 최빈값 다 지우기
                mode_list.append(list[i])
            elif n==most:
                mode_list.append(list[i])
        else:
            n=0
    if len(mode_list)==0:
        return list[1]
    elif len(mode_list)==1:
        return mode_list[0]
    else:
        return mode_list[1]


# 범위
def list_range(list):
    return list[-1]-list[0]

# round 함수를 사용하여 정수로 반올림 할 수 있다.
print(round(avg(num_list)))
print(mid(num_list))
print(mode(num_list))
print(list_range(num_list))

File: 백준/Silver/2108. 통계학/통계학.cc
#include <bits/stdc++.h>
using namespace std;

int avg(vector<int> list){
    double res = 0;
    for (int num : list){
        res += num;
    }
    return round(res / list.size());
}

int median(const vector<int>& list) {
    // 정렬이 되어있다는 가정 하에
    return list[list.size() / 2];
}

vector<int> mode(vector<int> list) {
    vector<int> modes;
    int cnt = 1;  // 현재 최빈값의 개수
    int temp = 1;  // 현재 진행중인 연속상태

    list.push_back(10000);  // 마지막까지 확인을 위해 임시로 넣어주기

    for (int i=1; i<list.size(); i++){
        if (list[i]==list[i-1]){  // 이전꺼랑 같다면
            temp++;
        }
        else if (temp > cnt){  // 연속 끊겼는데 최고기록을 넘었다면
            modes = {list[i-1]};
            cnt = temp;
            temp = 1;
        }
        else if (temp == cnt){  // 연속 끊겼는데 최고기록과 동일하다면
            modes.push_back(list[i-1]);
            temp = 1;
        }
        else temp = 1;  // 연속 끊겼는데 기록 못깼다면
    }
    return modes;
}

int list_range(const std::vector<int>& list) {
    return list.back() - list.front();
}



int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(0);

    int N, get_num;
    vector<int> nums;

    cin >> N;
    while(N--){
        cin >> get_num;
        nums.push_back(get_num);
    }
    sort(nums.begin(), nums.end());

    // 1. 평균
    cout << avg(nums) << '\n';
    // 2. 중앙값
    cout << median(nums) << '\n';

    // 3. 최빈값
    vector<int> modes = mode(nums);
    if (modes.size()==1) cout << modes[0] << '\n';
    else cout << modes[1] << '\n';

    // 4. 범위
    cout << list_range(nums) << '\n';

    
    return 0;
}